src/notifications/NotificationsManager.ts

```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
+    const notificationId = notificationsUtils.decodeNotificationId(
+      notificationIdEncoded,
+    )!;
+    const notificationKeyPath = [
+      ...this.notificationsManagerOutboxNotificationsDbPath,
+      idUtils.toBuffer(notificationId),
+    ];
+
+    const notificationDb = await this.db.withTransactionF(
+      async (tran) => (await tran.get<NotificationDB>(notificationKeyPath))!,
+    );
+    const signedNotification = await notificationsUtils.generateNotification(
+      {
+        notificationIdEncoded,
+        ...notificationDb,
+      },
+      this.keyRing.keyPair,
+    );
+    // The task id if a new task has been scheduled for a retry.
+    let newTaskId: TaskId | undefined;
+    try {
+      await this.nodeManager.withConnF(nodeId, async (connection) => {
+        const client = connection.getClient();
+        await client.methods.notificationsSend({
+          signedNotificationEncoded: signedNotification,
+        });
+      });
+    } catch (e) {
+      if (e instanceof errors.ErrorPolykeyRemote) {
+        this.logger.warn(
+          `Notification recipient ${nodesUtils.encodeNodeId(
+            nodeId,
+          )} responded with error: ${e.cause.description}`,
+        );
+        throw e;
+      } else {
+        this.logger.warn(
+          `Could not send to ${
+            notificationDb.data.type
+          } notification to ${nodesUtils.encodeNodeId(nodeId)}`,
+        );
+      }
+      // This will only happen on connection errors.
+      if (attempt < retries) {
+        // Recursively return inner task, so that the handler can process them.
+        const newTask = await this.taskManager.scheduleTask({
+          handlerId: this.sendNotificationHandlerId,
+          path: [this.sendNotificationHandlerId],
+          parameters: [
+            {
+              nodeIdEncoded,
+              notificationIdEncoded,
+              retries,
+              attempt: attempt + 1,
+            },
+          ],
+          delay: 1000 * 2 ** attempt,
+          lazy: false,
+        });
+        newTaskId = newTask.id;
+        return newTask;
+      } else {
+        throw e;
+      }
+    } finally {
+      await this.db.withTransactionF(async (tran) => {
+        // If a new task has been scheduled, set it in the map. If not, delete it.
+        if (newTaskId == null) {
+          this.outboxNotificationIdEncodedToTaskIdMap.delete(
+            notificationIdEncoded,
+          );
+          await tran.del(notificationKeyPath);
+          const counter = await tran.get<number>(
+            this.notificationsManagerOutboxNotificationsCounterDbPath,
+          );
+          await tran.put(
+            this.notificationsManagerOutboxNotificationsCounterDbPath,
+            counter != null ? counter - 1 : 0,
+          );
+        } else {
+          this.outboxNotificationIdEncodedToTaskIdMap.set(
+            notificationIdEncoded,
+            newTaskId,
+          );
+        }
+      });
+    }
```

Comment:
The logic flow of this whole try-catch-finally block seems a little hard to follow. So the logic is

1. Send succeeds - Preform any needed cleanup in database and state indicating success
2. Send fails and out of attempts - Clean up state indicating no more retries.
3. Send fails and more attempts - schedule new attempt and modify any needed state.

Right now this is spread across the try-catch-finally block. which is why it's confusing. I think we can flatten this out a little in just the try-catch sections. Doing a state update before returning or throwing.

Reply:
it's been made much simpler now with the removal of the NotificationID -> TaskID map